[RFC][WIP]analyze: add initial skelethon

igoropaniuk opened the pull request:
Add initial skelethon of the app for analyzing timestamps dumps. Currently
there are only stubs for different benchmark cases, but added
generic code for working with YAML file, parsing symbols (wrapper
for addr2line), and iterating through the set of timestamps.

Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>

Commit:
analyze: add initial skelethon

Add initial skelethon of the app for analyzing timestamps dumps. Currently
there are only stubs for different benchmark cases, but added
generic code for working with YAML file, parsing symbols (wrapper
for addr2line), and iterating through the set of timestamps.

Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>

jockebech commented on analyze.py:
```diff
@@ -0,0 +1,35 @@
+# path to all benchmark modules
+sys.path.append("analyze/")
+if sys.version_info[0] < 3:
+    raise "Python 3 should be used to run this script"
```
Would it be possible to rewrite the script so it works both on Python 2 and 3? If not, then I agree, it's better to aim for v3.

d3zd3z replied:
If you depend on python 3, please make the initial line in the script state that:

    #!/usr/bin/env python3

igoropaniuk replied:
@jbech-linaro you're right, it's better to do ASAP until there is no much code here

igoropaniuk replied:
@jbech-linaro decided to check if there are any python2 incompatibilities, found out that after installing back-ported `enum34` and `pathlib` everything works fine without any changes :)

jockebech commented on analyze.py:
```diff
@@ -0,0 +1,35 @@
+
+try:
+    if not (args.filename and Path(args.filename).is_file()):
+        raise RuntimeError("Wrong path to timestamp dump file")
```
What about changing the error message slightly? `"Cannot find file: {}.".format(args.filename)"`

igoropaniuk replied:
@jbech-linaro, thanks, will fix

jockebech commented on analyze/benchcases.py:
```diff
@@ -0,0 +1,29 @@
+    KMOD2TEE = 'kernel2tee'
+
+    def __str__(self):
+        return str(self.value)
```
There is no `self.value` in the class by default(?). I.e., you should probably initialize self.value in the constructor.

igoropaniuk replied:
Enum is not a regular python class, and here is a magic explained :) :

https://docs.python.org/3/library/enum.html#creating-an-enum

> class Color(Enum):
...     RED = 1
...     GREEN = 2
...     BLUE = 3
The class Color is an enumeration (or enum)
The attributes Color.RED, Color.GREEN, etc., are enumeration members (or enum members) and are functionally constants.
The enum members have names and values (the name of Color.RED is RED, the value of Color.BLUE is 3, etc.)
Even though we use the class syntax to create Enums, Enums are not normal Python classes (Enums have a custom metaclass that affects many aspects of both derived Enum classes and their instances (members).)

So in this case `self` will point to the member here, and enum members has preinitialized `.value` and `.name` attributes.

to understand the whole picture, just an example (some hypothetical example with `Planet` enumeration ):
```
>>> dir(Planet)
['EARTH', 'JUPITER', 'MARS', 'MERCURY', 'NEPTUNE', 'SATURN', 'URANUS', 'VENUS', '__class__', '__doc__', '__members__', '__module__']
>>> dir(Planet.EARTH)
['__class__', '__doc__', '__module__', 'name', 'surface_gravity', 'value']
```

jockebech replied:
Ah, I see and I didn't pay attention that the class inherited Enum, thanks for the explanation!

jockebech commented on analyze/config.py:
```diff
@@ -0,0 +1,20 @@
+from timestamp import Subsystem
+
+if sys.version_info[0] < 3:
+    raise "Python 3 should be used to run this script"
```
Same comment here as above regarding Python 2 vs 3.

jockebech commented on analyze/config.py:
```diff
@@ -0,0 +1,20 @@
+if __name__ == '__main__':
+    sys.exit("This source file should be used only as a module")
+
+subsystem_paths = {Subsystem.LIBTEEC:
```
These hard dependencies must be mentioned "somewhere", maybe a readme or when running `-h`.

igoropaniuk replied:
@jbech-linaro Sure, will add. I think I'll also add some additional app params (for each separate binary, or for the whole repo dir whatever) to make possible to overwrite it. Something like:

`./analyze --optee-working-dir=/home/user/dev/optee.repo/

jockebech commented on analyze/timestamp.py:
```diff
@@ -0,0 +1,128 @@
+
+        self.__parse_addr()
+
+    def __parse_addr(self):
```
Any specific reason why you are putting prefix `__foo` on non-standard Python functions in classes like this?

igoropaniuk replied:
@jbech-linaro 
I'm using double underscore for a name mangling , to avoid by error this "private" method invocations (as the method name will be mangled) outside the class (although PEP8 recommends to use single-leading-underscore convention instead for this purposes, and name mangling has a bit different purpose). I'll change it to single-leading-underscore, if we decided that single-leading-underscore is more preferable for such cases

Commit:
analyze: fixes after review

Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>

Commit:
readme: add initial readme file

Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>

igoropaniuk commented:
@jbech-linaro Added README.md for both `benchmark` and `analyze.py`

Commit:
analyze: fix addr2line resolve

Add other cosmetic fixes

Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>

jockebech commented on README.md:
```diff
@@ -1,2 +1,105 @@
+and implementation details of benchmark application.
+
+For a generic description of benchmark feature, its components in OP-TEE OS
+core, Linux TEE driver and libteec, please check [optee_benchmark.md].
```
[optee_benchmark.md] doesn't link, the reason is that further down you have written [optee_benchmark] (i.e., no `.md`).

jockebech commented on README.md:
```diff
@@ -1,2 +1,105 @@
+
+Benchmark application consists of two parts:
+- **Benchmark Client Application**, which is mainly used to generate latency
+  data. It should be invoked on a destination SoC.
```
s/It should be invoked on a destination SoC./You are supposed to run this on the device/?

jockebech commented on README.md:
```diff
@@ -1,2 +1,105 @@
+## 2.2 Analyze application
+`analyze.py` supports both Python 2 and 3. If you intend to use Python 2 or
+any other version >=3.4, you should install in advance back-ported
+`enum34` and `pathlib` packages using `pip` or `easy_install` package managers
```
Is it worth mention `pip install --user ...` to avoid "polluting" the system?

jockebech commented on README.md:
```diff
@@ -1,2 +1,105 @@
+- Address in subsystem, where timestamp was generated from.
+- Source file and line for address: `Timestamp` instance can automatically
+  resolve these values from symbol files, provided in `analyze/config.py`
+- Payload (not standartized). It could contain such values, like:
```
s/standartized/standardized/

jockebech commented on README.md:
```diff
@@ -1,2 +1,105 @@
+with `CFG_TEE_BENCHMARK` flag enabled. Check [optee_benchmark] for additional
+details.
+
+To collect benchmark data, you should invoke `benchmark`app and supply it with
```
nit: add space between `benchmark` and `app`.

jockebech commented on analyze/benchcases.py:
```diff
@@ -0,0 +1,44 @@
+
+
+def benchmark_client2tee(timestamp_parser):
+    # raise NotImplementedError(
```
Remove?

igoropaniuk replied:
sure ,and soon I will add a full implementation of `benchmark_client2tee`

jockebech commented on analyze/timestamp.py:
```diff
@@ -0,0 +1,150 @@
+
+    def lookup(self, addr):
+        try:
+            print(hex(addr))
```
Not a big deal, but you are mixing a bit from old (legacy) way of printing and the new (preferred) way using `.format{}`.

igoropaniuk replied:
@jbech-linaro particularly this print was added while debugging (and obviously I forgot to remove it). The same for next `print(info)` on line 54.

Other occurrences will be fixed, thanks!

igoropaniuk commented:
@jbech-linaro @d3zd3z thanks for reviewing! As I'm permanently adding new changes/altering existing code and keeping track of everything (to tell the truth, I'm already lost :) ), I think I will be better if I create a new PR from scratch  after addressing all your comments (without [WIP] and [RFC] tags)
